File: frontend/src/lib/oidc.ts
/**
 * OpenID Connect, authorization code with PKCE.
 *
 * Written rather than pulled in because the flow is small, and because the two
 * places it is easy to get wrong — verifying `state`, and never putting a token
 * where a stale one can outlive a sign-out — are easier to keep right in code
 * that is read than in configuration for a library that is not.
 *
 * There is no client secret. A single-page application cannot keep one, which
 * is the whole reason PKCE exists: the code verifier is generated per attempt,
 * never leaves this origin, and is what proves the code being redeemed belongs
 * to the browser that asked for it.
 *
 * ── This module holds no tokens ─────────────────────────────────────────────
 *
 * It used to. The browser redeemed the authorization code itself, kept the
 * access token in memory, and kept the refresh token in localStorage — because
 * surviving a page reload requires *something* persistent and every storage a
 * page can reach is readable by script on this origin. That was a documented
 * trade-off rather than an oversight, and it is now gone.
 *
 * The code is posted to `POST /api/v1/auth/callback` and the core redeems it.
 * What comes back is the same httpOnly session cookie a local password sign-in
 * gets. So this file builds an authorization URL, checks `state`, and hands the
 * result to the core; it never sees an access token, a refresh token, or an ID
 * token. There is nothing here for a cross-site scripting flaw to steal that it
 * could not already do by calling the API directly.
 *
 * The nonce is generated here and sent to the core alongside the code, because
 * the core is the half that must check it. A browser verifying its own nonce
 * proves nothing.
 */

export interface AuthConfig {
  // 'password' when only local accounts are configured, 'oidc' when single
  // sign-on is offered as well. There is no anonymous mode.
  mode: 'oidc' | 'password'
  password_login?: boolean
  issuer?: string
  client_id?: string
  scopes?: string[]
  audience?: string
}

interface ProviderMetadata {
  authorization_endpoint: string
  token_endpoint: string
  end_session_endpoint?: string
}

const PENDING_KEY = 'certpilot.auth_pending'
const RETURN_KEY = 'certpilot.return_to'

/**
 * The key an older build kept a refresh token under.
 *
 * Removed on load rather than merely stopped being written. An upgrade
 * otherwise leaves a live refresh token sitting in localStorage indefinitely,
 * belonging to a flow nothing uses any more — the exact object this change
 * exists to get rid of, preserved by the change that removed it.
 */
const LEGACY_REFRESH_KEY = 'certpilot.refresh_token'

/** Marks that a sign-in has completed in this browser. Carries no credential. */
const SIGNED_IN_KEY = 'certpilot.signed_in'
try {
  localStorage.removeItem(LEGACY_REFRESH_KEY)
} catch {
  // Storage can be unavailable — a private window, or a browser configured to
  // block it. Nothing here depends on the removal succeeding.
}

// ── Small crypto helpers ────────────────────────────────────────────────────

function randomString(bytes = 32): string {
  const buf = new Uint8Array(bytes)
  crypto.getRandomValues(buf)
  return base64url(buf)
}

function base64url(bytes: Uint8Array): string {
  let binary = ''
  for (const b of bytes) binary += String.fromCharCode(b)
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '')
}

async function challengeFor(verifier: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(verifier))
  return base64url(new Uint8Array(digest))
}

// ── Discovery ───────────────────────────────────────────────────────────────

let cachedConfig: AuthConfig | null = null
let cachedMetadata: ProviderMetadata | null = null

/**
 * Reads how this instance authenticates, from the core rather than from the
 * build.
 *
 * A frontend carrying its own copy of the issuer can disagree with the API it
 * talks to, and that disagreement presents as a login that appears to succeed
 * followed by 401 on every request — which reads as a broken server rather
 * than a mismatched setting.
 */
export async function loadAuthConfig(): Promise<AuthConfig> {
  if (cachedConfig) return cachedConfig
  const response = await fetch('/api/v1/auth/config')
  if (!response.ok) {
    throw new Error(
      `CertPilot could not be asked how to sign in (HTTP ${response.status}). ` +
        'The API may be unreachable.',
    )
  }
  cachedConfig = (await response.json()) as AuthConfig
  return cachedConfig
}

async function providerMetadata(issuer: string): Promise<ProviderMetadata> {
  if (cachedMetadata) return cachedMetadata
  const url = `${issuer.replace(/\/$/, '')}/.well-known/openid-configuration`
  const response = await fetch(url)
  if (!response.ok) {
    throw new Error(
      `The identity provider at ${issuer} did not return its OpenID configuration ` +
        `(HTTP ${response.status}). Check auth.issuer in the core's configuration.`,
    )
  }
  cachedMetadata = (await response.json()) as ProviderMetadata
  return cachedMetadata
}

function redirectUri(): string {
  return `${window.location.origin}/auth/callback`
}

// ── Sign-in ─────────────────────────────────────────────────────────────────

/** Sends the browser to the identity provider. Does not return. */
export async function beginSignIn(returnTo?: string): Promise<void> {
  const config = await loadAuthConfig()
  if (config.mode !== 'oidc' || !config.issuer || !config.client_id) {
    throw new Error('This CertPilot instance is not configured for single sign-on.')
  }

  const metadata = await providerMetadata(config.issuer)
  const verifier = randomString()
  const state = randomString(16)
  const nonce = randomString(16)

  // sessionStorage, not localStorage: this is per-attempt and per-tab, and a
  // verifier surviving in another tab is a verifier that can be replayed.
  sessionStorage.setItem(PENDING_KEY, JSON.stringify({ verifier, state, nonce }))
  if (returnTo) sessionStorage.setItem(RETURN_KEY, returnTo)

  const params = new URLSearchParams({
    response_type: 'code',
    client_id: config.client_id,
    redirect_uri: redirectUri(),
    scope: (config.scopes ?? ['openid', 'profile', 'email']).join(' '),
    state,
    nonce,
    code_challenge: await challengeFor(verifier),
    code_challenge_method: 'S256',
  })
  if (config.audience) params.set('audience', config.audience)

  window.location.assign(`${metadata.authorization_endpoint}?${params.toString()}`)
}

/**
 * Completes the flow after the provider redirects back.
 *
 * Returns where the user was trying to go, so the caller can put them back
 * there rather than always landing them on the dashboard.
 */
export async function completeSignIn(search: string): Promise<string> {
  const params = new URLSearchParams(search)

  const error = params.get('error')
  if (error) {
    throw new Error(
      `${params.get('error_description') ?? error}. ` +
        'This came from the identity provider, not from CertPilot.',
    )
  }

  const code = params.get('code')
  const returnedState = params.get('state')
  const pendingRaw = sessionStorage.getItem(PENDING_KEY)
  sessionStorage.removeItem(PENDING_KEY)

  if (!code || !pendingRaw) {
    throw new Error('This sign-in did not start here. Begin again from the login page.')
  }

  const pending = JSON.parse(pendingRaw) as { verifier: string; state: string; nonce: string }

  // The check that makes the redirect safe. Without it, an attacker can hand
  // somebody a callback URL carrying their own authorization code and have the
  // victim's browser sign in as the attacker — the session then looks entirely
  // legitimate to the person using it.
  if (returnedState !== pending.state) {
    throw new Error(
      'The sign-in response did not match the request that started it, so it was refused.',
    )
  }

  // The core redeems the code, not this page. What comes back is an httpOnly
  // session cookie — the same one a local password sign-in gets — so nothing
  // token-shaped ever reaches this origin's storage or this module's memory.
  //
  // The nonce goes with it. It was generated here, and the core is the half
  // that checks the ID token carries the same one; a browser verifying its own
  // nonce proves nothing.
  const response = await fetch('/api/v1/auth/callback', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    credentials: 'same-origin',
    body: JSON.stringify({
      code,
      code_verifier: pending.verifier,
      redirect_uri: redirectUri(),
      nonce: pending.nonce,
    }),
  })

  if (!response.ok) {
    const body = (await response.json().catch(() => ({}))) as { error?: string }
    throw new Error(
      body.error ??
        `This sign-in could not be completed (HTTP ${response.status}). ` +
          'The most common cause is a redirect URI that is not registered for this client.',
    )
  }

  rememberSignIn()

  const returnTo = sessionStorage.getItem(RETURN_KEY) ?? '/'
  sessionStorage.removeItem(RETURN_KEY)
  return returnTo
}

// ── Keeping the session alive ───────────────────────────────────────────────

/**
 * Whether a previous session might still be usable without interaction.
 *
 * The session cookie is httpOnly, so this page cannot see it. The honest answer
 * is therefore "ask the API", which the auth store does on start-up — and the
 * only thing this can report is whether this browser has ever completed a
 * sign-in here.
 *
 * It exists to keep one message truthful. "Your session ended" shown to a
 * first-time visitor who has simply arrived at the URL is wrong and alarming;
 * the same message after a session really has expired is exactly right.
 */
export function hasResumableSession(): boolean {
  try {
    return localStorage.getItem(SIGNED_IN_KEY) === '1'
  } catch {
    return false
  }
}

/** Records that a sign-in has completed here, for hasResumableSession. */
export function rememberSignIn(): void {
  try {
    localStorage.setItem(SIGNED_IN_KEY, '1')
  } catch {
    // Not worth failing a sign-in over.
  }
}

/**
 * Forgets local sign-in state.
 *
 * There are no tokens left to clear — the session lives in an httpOnly cookie
 * the core sets and clears. This drops only the marker above, so that the next
 * visit is treated as a first one rather than as an expired session.
 */
export function clearTokens(): void {
  try {
    localStorage.removeItem(SIGNED_IN_KEY)
    localStorage.removeItem(LEGACY_REFRESH_KEY)
  } catch {
    // As above.
  }
}

/**
 * Ends the session here, and at the provider when it supports it.
 *
 * Clearing only the local tokens leaves the provider's own session cookie
 * intact, so the next sign-in completes silently and looks to the operator as
 * though signing out did nothing.
 */
export async function signOut(): Promise<void> {
  clearTokens()
  try {
    const config = await loadAuthConfig()
    if (config.mode !== 'oidc' || !config.issuer) return
    const metadata = await providerMetadata(config.issuer)
    if (!metadata.end_session_endpoint) return

    const params = new URLSearchParams({
      client_id: config.client_id!,
      post_logout_redirect_uri: `${window.location.origin}/login`,
    })
    window.location.assign(`${metadata.end_session_endpoint}?${params.toString()}`)
  } catch {
    // Signing out must never fail in a way that leaves somebody signed in.
    window.location.assign('/login')
  }
}

File: core/api/oidc.go
package api

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/certpilot/certpilot/core/server/middleware"
	"github.com/gin-gonic/gin"
)

// The core redeems the authorization code, not the browser.
//
// This is a backend-for-frontend, and it exists to answer one question: where
// does the credential that survives a page reload live? A single-page
// application that redeems the code itself gets a refresh token it has nowhere
// safe to put — every storage a page can reach is readable by script on that
// origin — so CertPilot kept it in localStorage and documented the trade-off.
//
// Once browser sessions became rows with an httpOnly cookie, that trade-off
// stopped being necessary. The browser now sends the code here, gets a session
// cookie back, and never handles a token at all. The refresh token is not
// stored anywhere: CertPilot's own session is the durable credential, and
// keeping a second one would mean holding a provider secret at rest to
// duplicate something the sessions table already does.
//
// What that costs, stated plainly: a federated session outlives revocation at
// the identity provider until it expires. CertPilot's answer to "this person
// should no longer have access" is suspending the account, which is checked on
// every request and ends every session immediately — the provider says who you
// are, CertPilot says what you may do.

// oidcExchangeTimeout bounds the call to the provider.
//
// A sign-in that hangs is a sign-in that fails; leaving the request open longer
// than this just holds a connection while somebody stares at a spinner.
const oidcExchangeTimeout = 15 * time.Second

// discoveryTTL is how long a provider's metadata is reused.
//
// Endpoints move about once in the lifetime of a deployment, so this is long.
// It is cached at all because the alternative is a well-known fetch on every
// sign-in, which turns the provider's availability into a hard dependency of
// each individual login rather than of the first one after a restart.
const discoveryTTL = time.Hour

type providerMetadata struct {
	TokenEndpoint string `json:"token_endpoint"`
	Issuer        string `json:"issuer"`
}

type discoveryCache struct {
	mu        sync.Mutex
	metadata  providerMetadata
	fetchedAt time.Time
}

var oidcDiscovery discoveryCache

// CallbackInput is what the browser brings back from the identity provider.
type CallbackInput struct {
	Code         string `json:"code" binding:"required"`
	CodeVerifier string `json:"code_verifier" binding:"required"`
	RedirectURI  string `json:"redirect_uri" binding:"required"`
	// Nonce is the value the browser put in the authorization request. It is
	// sent back so the core can check the ID token carries the same one — the
	// browser cannot be trusted to check it itself, because the whole point of
	// moving this here is that the browser is the untrusted half.
	Nonce string `json:"nonce" binding:"required"`
}

// Callback handles POST /api/v1/auth/callback.
//
// Public, because a sign-in cannot require being signed in. Every failure
// returns the same shape of message: the reasons are useful to an operator
// reading the log and useful to somebody probing the endpoint, so only the
// former gets them.
func (h *SessionHandler) Callback(c *gin.Context) {
	if h.auth.Issuer == "" || h.auth.ClientID == "" {
		c.JSON(http.StatusNotFound, gin.H{
			"error": "this instance is not configured for single sign-on; sign in with an account instead",
		})
		return
	}

	var input CallbackInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "this sign-in response was incomplete; begin again from the sign-in page",
		})
		return
	}

	ctx, cancel := context.WithTimeout(c.Request.Context(), oidcExchangeTimeout)
	defer cancel()

	metadata, err := h.providerMetadata(ctx)
	if err != nil {
		slog.Error("could not read the identity provider's metadata", "error", err, "issuer", h.auth.Issuer)
		c.JSON(http.StatusBadGateway, gin.H{
			"error": "CertPilot could not reach the identity provider to complete this sign-in",
		})
		return
	}

	idToken, err := h.exchangeCode(ctx, metadata.TokenEndpoint, input)
	if err != nil {
		slog.Warn("the authorization code exchange failed", "error", err, "client_ip", c.ClientIP())
		c.JSON(http.StatusUnauthorized, gin.H{
			"error": "the identity provider refused this sign-in; begin again from the sign-in page",
		})
		return
	}

	claims, err := h.authenticator.VerifyIDToken(ctx, idToken, h.auth.ClientID, input.Nonce)
	if err != nil {
		// Logged in full and reported in outline. The reason names which part
		// of a forgery to correct.
		slog.Warn("an ID token was refused", "error", err, "client_ip", c.ClientIP())
		c.JSON(http.StatusUnauthorized, gin.H{
			"error": "the identity provider's response could not be verified, so it was refused",
		})
		return
	}

	user, err := h.authenticator.ResolveIdentity(ctx, claims)
	if err != nil {
		// Fail closed, exactly as the request path does. A role that cannot be
		// established is not a role.
		slog.Error("could not resolve a federated identity", "error", err, "subject", claims.Subject)
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "your identity could not be resolved against CertPilot's user directory, " +
				"so no role could be established",
		})
		return
	}

	if !user.IsActive() {
		h.audit(c, user.Subject, "auth.login_failed", map[string]any{
			"reason": "account suspended", "email": user.Email, "method": "oidc",
		})
		c.JSON(http.StatusForbidden, gin.H{
			"error": "this account is suspended in CertPilot; the identity provider still accepts it",
		})
		return
	}

	raw, hash, err := middleware.NewSessionToken()
	if err != nil {
		slog.Error("could not generate a session token", "error", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "sign-in could not be completed"})
		return
	}

	expires := time.Now().Add(middleware.SessionLifetime)
	if _, err := h.store.CreateSession(ctx, user.ID, hash, expires,
		c.Request.UserAgent(), c.ClientIP()); err != nil {
		slog.Error("could not create a session", "error", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "sign-in could not be completed"})
		return
	}

	middleware.SetSessionCookie(c, raw, expires)
	h.audit(c, user.Subject, "auth.login", map[string]any{"email": user.Email, "method": "oidc"})

	c.JSON(http.StatusOK, MeResponse{
		Subject:     user.Subject,
		Email:       user.Email,
		DisplayName: user.DisplayName,
		Role:        user.Role,
		AuthMethod:  middleware.AuthMethodSession,
		UserID:      user.ID,
		RoleSource:  user.RoleSource,
		// A federated account has no CertPilot password to change, but the
		// field is carried rather than omitted so that one response shape means
		// one thing everywhere.
		MustChangePassword: user.MustChangePassword,
	})
}

// exchangeCode redeems the authorization code and returns the ID token.
//
// No client secret is sent, and none is configured: this is a public client and
// PKCE is what proves the code belongs to the browser that asked for it. The
// access and refresh tokens in the response are deliberately dropped on the
// floor — see the note at the top of this file.
func (h *SessionHandler) exchangeCode(ctx context.Context, tokenEndpoint string, input CallbackInput) (string, error) {
	form := url.Values{
		"grant_type":    {"authorization_code"},
		"code":          {input.Code},
		"redirect_uri":  {input.RedirectURI},
		"client_id":     {h.auth.ClientID},
		"code_verifier": {input.CodeVerifier},
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, tokenEndpoint,
		strings.NewReader(form.Encode()))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	// Bounded: the response is a small JSON document, and an endpoint that
	// streams gigabytes at the core should not be able to.
	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return "", err
	}
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("token endpoint returned HTTP %d: %s", resp.StatusCode, truncate(string(body), 200))
	}

	var tokens struct {
		IDToken string `json:"id_token"`
	}
	if err := json.Unmarshal(body, &tokens); err != nil {
		return "", fmt.Errorf("token endpoint returned something that is not JSON: %w", err)
	}
	if tokens.IDToken == "" {
		return "", fmt.Errorf("token endpoint returned no id_token; check that the openid scope is requested")
	}
	return tokens.IDToken, nil
}

// providerMetadata reads and caches the provider's discovery document.
func (h *SessionHandler) providerMetadata(ctx context.Context) (providerMetadata, error) {
	oidcDiscovery.mu.Lock()
	defer oidcDiscovery.mu.Unlock()

	if oidcDiscovery.metadata.TokenEndpoint != "" &&
		oidcDiscovery.metadata.Issuer == h.auth.Issuer &&
		time.Since(oidcDiscovery.fetchedAt) < discoveryTTL {
		return oidcDiscovery.metadata, nil
	}

	endpoint := strings.TrimSuffix(h.auth.Issuer, "/") + "/.well-known/openid-configuration"
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return providerMetadata{}, err
	}
	req.Header.Set("Accept", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return providerMetadata{}, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return providerMetadata{}, fmt.Errorf("%s returned HTTP %d", endpoint, resp.StatusCode)
	}

	var metadata providerMetadata
	if err := json.NewDecoder(io.LimitReader(resp.Body, 1<<20)).Decode(&metadata); err != nil {
		return providerMetadata{}, err
	}
	if metadata.TokenEndpoint == "" {
		return providerMetadata{}, fmt.Errorf("%s names no token_endpoint", endpoint)
	}

	// The issuer in the document must be the one that was asked, or the
	// discovery has been redirected somewhere else — which is how a
	// misconfigured proxy turns into an authentication bypass.
	if metadata.Issuer != "" && metadata.Issuer != h.auth.Issuer {
		return providerMetadata{}, fmt.Errorf(
			"discovery at %s describes issuer %q, not the configured %q",
			endpoint, metadata.Issuer, h.auth.Issuer)
	}
	metadata.Issuer = h.auth.Issuer

	oidcDiscovery.metadata = metadata
	oidcDiscovery.fetchedAt = time.Now()
	return metadata, nil
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "…"
}
